_drafts/STMViewer.md

```diff
@@ -0,0 +1,161 @@
+
+2. Trace Viewer - a module that reads and parses SWO (Serial Wire Viewer) data using an ST-Link. To use it, you need to connect SWDIO, SWCLK, SWO, and GND. This tool operates synchronously, meaning it visualizes data points at the rate they are produced on the target. A data point can be either a special single-byte flag used to create digital plots or any value that fits within up to 4 bytes, as in the case of 'analog' plots. It incurs minimal performance overhead, involving just a single register write per plot point, as serialization and timestamping are handled by the ITM peripheral. Trace Viewer can be employed for profiling specific parts of the codebase or visualizing signals that are too fast for the asynchronous Variable Viewer. Its main limitations are the maximum allowed SWO pin baudrate of the ST-Link and the fact that the ITM peripheral is supported only on Cortex M3/M4/M7/M33 cores. 
+
+
+## I'm already using STMStudio/CubeMonitor, why create a new tool?
+
+Let's begin with the most commonly asked question: Why bother creating a tool when existing options are provided by ST? There are several reasons:
+
+1. STMStudio is deprecated, and while it was a useful tool, it became challenging to work with due to persistent bugs. It also lacks support for mangled C++ names and is exclusively available for Windows.
+
+2. CubeMonitor, on the other hand, is often considered too much overhead for debugging purposes. It's primarily designed for creating dashboards, and during debugging, the focus tends to be on functionality rather than aesthetics. Manipulating plot data and writing values to logged variables can be inconvenient. Furthermore, setting up a basic logger configuration can require some time to become familiar with NodeRED's setup.
+
+3. Last but not least, none of the available hobby-level software tools include a trace data visualizer. In my opinion, STMViewer's trace data visualizer is the most intriguing module, enabling users to visualize variables synchronously and profile functions or interrupts within minutes.
+
+## Variable Viewer
+
+<!-- TODO GIF ? -->
+
+As mentioned earlier, Variable Viewer is an asynchronous module that samples values from predefined RAM addresses on your target. It can be particularly useful when you want to visualize multiple, relatively slow signals (with a maximum sampling rate of around 1 kHz). The RAM addresses are read from the project's *.elf file, which is parsed using GDB. This is essentially the same approach used by STMStudio and CubeMonitor. 
+
+For a quick start please refer to: [Variable Viewer](https://github.com/klonyyy/STMViewer#variable-viewer-1)
+
+### PID controller
+Let's see it in action on a step response of a classic cascaded PID from a servo drive:
```

Comment:
Is the source for this available anywhere?

Reply:
I can prepare a simplified PID implementation that was used for both controllers, but apart from that I think it would be too much to describe in a single usage example.